### PR TITLE
feat: extend types, improve ci workflow

### PR DESCRIPTION
**Pull request by e965**
1) `.editorconfig` file added
2) repository moved from yarn to npm (following the example of 8base cli, where yarn is not used either) - simplifying the maintanance of the project (maybe)
3) improving CI/CD processes
4) `ResolverFunction` and `WebhookFunction` utility types added

This last improvement will allow custom functions to be declared not like this -

```ts
const eightBaseCustomFunction = async (
  event: FunctionEvent<CustomFunctuionArgs>,
  ctx: FunctionContext,
): Promise<FunctionResponse<CustomFunctuionResponse>> => {
  ...
}
```

But more simply, like this - 

```ts
const eightBaseCustomFunction: ResolverFunction<CustomFunctuionArgs, CustomFunctuionResponse> = async (
  event, ctx,
) => {
  ..
}
```

### CODE REVIEW

**Comment by e965** [comment]
@s1owjke take a look please 🙂

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,9 @@
+root = true
+
+[*]
+indent_style = space
+indent_size = 2
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
```
Comment (s1owjke) [inline .editorconfig:9]: For consistency between projects let's use this one

```suggestion
root = true

[*]
insert_final_newline = true
trim_trailing_whitespace = true
indent_style = space
indent_size = 2
```

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -1,28 +1,40 @@
...
+
+concurrency:
+  group: ci-publish
+  cancel-in-progress: true
 
 jobs:
   publish:
     runs-on: ubuntu-latest
 
+    permissions:
```
Comment (s1owjke) [inline .github/workflows/publish.yml:15]: Why should we define `permissions` explicitly?

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -1,28 +1,40 @@
...
-        run: yarn typecheck
+      - name: Build project
+        run: npm run build
 
       - name: Publish
-        run: npm publish --access public
-
-        env:
-          NODE_AUTH_TOKEN: ${{ secrets.NPM_AUTH_TOKEN_ORG }}
+        uses: JS-DevTools/npm-publish@v2
```
Comment (s1owjke) [inline .github/workflows/publish.yml:36]: Let's revert this, for consistency and reliability

File: .npmignore [hunk 4]
```diff
@@ -1 +1,2 @@
 yarn.lock
+*.tgz
```
Comment (s1owjke) [inline .npmignore:2]: Do we use `.tgz` extension? 

Also, let's ignore sources and Typescript configs, as we don't need them in npm registry

```suggestion
/src
tsconfig.json
```

File: package.json [hunk 5]
```diff
@@ -1,17 +1,26 @@
...
     "typecheck": "tsc --noEmit",
     "postversion": "git push --follow-tags"
   },
   "dependencies": {
     "@graphql-typed-document-node/core": "^3.2.0",
-    "graphql": "^15.8.0"
+    "graphql": "^16.6.0"
+  },
+  "devDependencies": {
+    "typescript": "^5.0.4"
```
Comment (s1owjke) [inline package.json:24]: Almost everywhere we currently use this version

```suggestion
    "typescript": "^4.3.5"
```

File: src/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export type AnyObject = Record<string, any>;
+
+export type Maybe<T> = T | null;
```
Comment (s1owjke) [inline src/helpers.ts:3]: In server it represents [flow's](https://flow.org/en/docs/types/maybe/) `Maybe` type and includes `undefined`. To make it less confusing let's remove it and define explicitly in types

```suggestion
export type Maybe<T> = null | undefined | T;
```

File: src/index.ts [hunk 7]
```diff
@@ -135,6 +131,7 @@ export type AfterDeleteTriggerFunctionEvent<
   ExtendObjectT = AnyObject
 > = {
   data: DataT & IDBObject;
+  originalData: OriginalDataT;
```
Comment (s1owjke) [inline src/index.ts:134]: @Rusty-Beard as I remember we don't have `originalData` here, right?

File: src/index.ts [hunk 8]
```diff
@@ -150,3 +147,17 @@ export type WebhookResponse = {
   headers?: HttpHeaders;
   body?: string;
 };
+
+/** The most common case for resolver functions  */
+
+export type ResolverFunction<EventData = {}, ResultData = {}> = (
```
Comment (s1owjke) [inline src/index.ts:153]: Good idea, @Rusty-Beard maybe define the same for every type?

File: tsconfig.json [hunk 9]
```diff
@@ -1,19 +1,19 @@
 {
   "compilerOptions": {
-    "diagnostics": true,
+    "diagnostics": false,
```
Comment (s1owjke) [inline tsconfig.json:3]:
```suggestion
    "diagnostics": true,
```

File: src/index.ts [hunk 10]
```diff
@@ -1,46 +1,48 @@
...
+>(
+  name: string,
+  args?: ArgsT,
+  options?: { waitForResponse: boolean; checkPermissions?: boolean }
+) => Promise<ResponseT>;
+
 export type FunctionContext = {
   api: {
     gqlRequest: GqlRequest;
+    /** @deprecated Method is not implemented */
```
Comment (s1owjke) [inline src/index.ts:37]: I think better to remove it completely, actually it's not deprecated as you see from message.

**Review by s1owjke — COMMENTED** [review]
Thanks for the suggestions, let's fix the comments and release

**Review by s1owjke — COMMENTED** [review]
Thanks for the suggestions, let's fix the comments and release

**Review by s1owjke — COMMENTED** [review]
Thanks for the suggestions, let's fix the comments and release

File: src/index.ts (hunk 7 above)
Comment (e965) [reply src/index.ts:134]: It is defined in the generic, so I added this line (just in case)

File: .github/workflows/publish.yml (hunk 2 above)
Comment (e965) [reply .github/workflows/publish.yml:15]: This is necessary for the "provenance" option to work correctly
https://docs.npmjs.com/generating-provenance-statements

**Comment by e965** [comment]
@s1owjke @Rusty-Beard commit fixes based on your reviews 🙏

**Review by s1owjke — APPROVED** [review]

**Pull request merged**